storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+}
+
+/**
+ * @return Selectivity of a greater or equals predicate.
+ */
+inline Selectivity getSelectivity_GE(const TypedValue &literal,
+                                     const TypedValue &min,
+                                     const TypedValue &max,
+                                     const UncheckedComparator *less_comparator,
+                                     const UncheckedComparator *equals_comparator) {
+  if (less_comparator->compareTypedValues(literal, min) ||
+      equals_comparator->compareTypedValues(literal, min)) {
+    return Selectivity::kAll;
+  } else if (less_comparator->compareTypedValues(max, literal)) {
+    return Selectivity::kNone;
+  }
+  return Selectivity::kSome;
+}
+
+Selectivity getSelectivity(const TypedValue &literal,
+                           const ComparisonID comparison,
+                           const TypedValue &min,
+                           const TypedValue &max,
+                           const UncheckedComparator *less_comparator,
+                           const UncheckedComparator *equals_comparator) {
+  switch (comparison) {
+    case ComparisonID::kEqual:
+      return getSelectivity_E(literal, min, max, less_comparator);
+    case ComparisonID::kLess:
+      return getSelectivity_L(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kLessOrEqual:
+      return getSelectivity_LE(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kGreater:
+      return getSelectivity_G(literal, min, max, less_comparator, equals_comparator);
+    case ComparisonID::kGreaterOrEqual:
+      return getSelectivity_GE(literal, min, max, less_comparator, equals_comparator);
+    default:
+      return Selectivity::kUnknown;
+  }
+}
+
+SMAPredicate* SMAPredicate::ExtractSMAPredicate(const ComparisonPredicate &predicate) {
+  if (predicate.getLeftOperand().hasStaticValue()) {
+    DLOG_IF(FATAL, predicate.getRightOperand().getDataSource() != Scalar::kAttribute);
+    return new SMAPredicate(
+         static_cast<const ScalarAttribute&>(predicate.getRightOperand()).getAttribute().getID(),
+         flipComparisonID(predicate.getComparison().getComparisonID()),
+         predicate.getLeftOperand().getStaticValue().makeReferenceToThis());
+  } else {
+    DLOG_IF(FATAL, predicate.getLeftOperand().getDataSource() != Scalar::kAttribute);
+    return new SMAPredicate(
+        static_cast<const ScalarAttribute&>(predicate.getLeftOperand()).getAttribute().getID(),
+        predicate.getComparison().getComparisonID(),
+        predicate.getRightOperand().getStaticValue().makeReferenceToThis());
+  }
+}
+
+/**
+ * @brief Summation will promote lower precision types to higher precision types.
+ *
+ * @param type The type id of the type we wish to sum over.
+ * @return The higher precision sum typeid. Returns kNullType to indicate that the
+ *         given type cannot be summed (in lieu of a better sigil).
+ */
+inline TypeID sumType(TypeID type) {
```

Comment:
A better name should be ``getSumType`` or ``getTypeForSum``. SQL uses "aggregate" instead of sum, so the term aggregation is even better - but you can choose whatever you prefer.